### PR TITLE
Feat/XRPL connector v2

### PR DESCRIPTION
**Pull request by mlguys**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
- Implemented CLOB api data source for XRPL Gateway connector
- Added tests to test the api data source
- Enable HB client to run strategies that support CLOB DEX connector, such as Pure Market Making strategy

**Tests performed by the developer**:
- Added wallets and connected to XRPL chain using `gateway`
- Checked balances on XRPL chain
- Created and run PMM strategy on XRPL DEX, using `USD/VND` and `XRP/VND` markets

**Tips for QA testing**:
- Bounty connected to this PR: https://github.com/hummingbot/gateway/issues/103
- Gateway PR connected to this PR: https://github.com/hummingbot/gateway/pull/128
- Proposal: https://snapshot.org/#/hbot-ip.eth/proposal/0x07b027fc420274d26add346ed65a2f7e3e5a662bd5317d2ddb2dd02562a7b2d2
- For end-to-end tests, please contact me for getting test wallets on testnet
- Link to trade USD/VND and XRP/VND on XRP DEX on testnet:
```
- USD/VND testnet book:
https://sologenic.org/trade?n[etwork=test](https://sologenic.org/trade?market=USD%2Brh8LssQyeBdEXk7Zv86HxHrx8k2R2DBUrx%2FVND%2Brh8LssQyeBdEXk7Zv86HxHrx8k2R2DBUrx)net&market=USD%2Brh8LssQyeBdEXk7Zv86HxHrx8k2R2DBUrx%2FVND%2Brh8LssQyeBdEXk7Zv86HxHrx8k2R2DBUrx

- XRP/VND testnet book:
https://sologenic.org/trade?n[etwork=](https://sologenic.org/trade?market=USD%2Brh8LssQyeBdEXk7Zv86HxHrx8k2R2DBUrx%2FVND%2Brh8LssQyeBdEXk7Zv86HxHrx8k2R2DBUrx)testnet&market=USD%2Brh8LssQyeBdEXk7Zv86HxHrx8k2R2DBUrx%2FVND%2Brh8LssQyeBdEXk7Zv86HxHrx8k2R2DBUrx
```
- To trade on other markets, please contact me for how to do that as you will need to update connector config

### CODE REVIEW

File: hummingbot/connector/gateway/clob_spot/data_sources/xrpl/xrpl_utils.py
Comment (cardosofede) [inline hummingbot/connector/gateway/clob_spot/data_sources/xrpl/xrpl_utils.py:1]: @mlguys this file is empty

File: hummingbot/connector/gateway/clob_spot/data_sources/xrpl/xrpl_api_data_source.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+from hummingbot.core.data_type.in_flight_order import InFlightOrder, OrderUpdate, TradeUpdate
+
+# from hummingbot.core.data_type.order_book_message import OrderBookMessage, OrderBookMessageType
+from hummingbot.core.data_type.trade_fee import MakerTakerExchangeFeeRates, TokenAmount, TradeFeeBase, TradeFeeSchema
+from hummingbot.core.event.events import MarketEvent
+from hummingbot.core.gateway.gateway_http_client import GatewayHttpClient
+from hummingbot.core.utils.async_utils import safe_gather
+from hummingbot.core.utils.tracking_nonce import NonceCreator
+
+# from hummingbot.core.web_assistant.connections.data_types import RESTMethod, WSJSONRequest
```
Comment (cardosofede) [inline hummingbot/connector/gateway/clob_spot/data_sources/xrpl/xrpl_api_data_source.py:37]: remove the commented imports

Comment (mlguys) [reply hummingbot/connector/gateway/clob_spot/data_sources/xrpl/xrpl_api_data_source.py]: Removed

File: hummingbot/connector/gateway/clob_spot/data_sources/xrpl/xrpl_utils.py
Comment (mlguys) [reply hummingbot/connector/gateway/clob_spot/data_sources/xrpl/xrpl_utils.py:1]: Removed

File: hummingbot/connector/gateway/gateway_in_flight_order.py [hunk 2]
```diff
@@ -9,7 +9,7 @@
 from hummingbot.core.data_type.in_flight_order import InFlightOrder, OrderState, OrderUpdate, TradeUpdate
 
 GET_GATEWAY_EX_ORDER_ID_TIMEOUT = 30  # seconds
-GET_GATEWAY_TX_HASH = 1  # seconds
+GET_GATEWAY_TX_HASH = 10  # seconds
```
[the remark this answers is not in the document]
Comment (mlguys) [reply hummingbot/connector/gateway/gateway_in_flight_order.py:12]: Hey @fengtality, I increased this as transactions on the Ripple network take between 3-5 seconds to process.

**Comment by nikspz** [comment]
@mlguys Could you please resolve branch conflicts?

**Comment by nikspz** [comment]
hi @mlguys  Could you please resolve branch conflicts?

**Comment by mlguys** [comment]
Hi @nikspz, sorry for the delay, conflicts resolved.

**Comment by nikspz** [comment]
@mlguys Could you please check unit tests failed (test coverage)

**Review by cardosofede — APPROVED** [review]
LGTM!

**Review by rapcmia — APPROVED** [review]
LGTM:
- Setup gateway connection with https://github.com/hummingbot/gateway/pull/128
- Setup wallet using XRP testnet `wss://s.altnet.rippletest.net:51233`
- Setup simple pmm on USD/VND
    image
     - Orders are created and canceled successfully on set refresh time
     - Orders match on client and exchange
        image
     - Filled trade event
        - Login 2nd account to take the order of 1st account
        - Trade data on client and exchange match, excluding fees on `history` command
- Build docker image and run tests via docker compose ✅

**Review by fengtality — APPROVED** [review]
LGTM

**Comment by rapcmia** [comment]
Merged to development and part of release version 1.23.0

**Comment by mlguys** [comment]
Thanks for the reminder @david-hummingbot, working on it!

**Pull request merged**